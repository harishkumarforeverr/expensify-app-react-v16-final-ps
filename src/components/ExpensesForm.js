import React from 'react';
import "react-dates/initialize";
import {SingleDatePicker} from "react-dates";
import moment from "moment";
 
export default class ExpensesForm extends React.Component{
    constructor(props){
        super(props)
        this.state={
            description:props.description?props.description:"",
            note:props.note?props.note:"",
            amount:props.amount?parseFloat(props.amount,10):"",
            createdAt:props.createdAt?moment(props.createdAt):moment(),
            error:undefined,
            focused:false
        }
    }
  onDescriptionChange=(e)=>{
    const description=e.target.value;
     this.setState(()=>({ 
         description
       }))
  }      
  onNoteChange=(e)=>{
    const note=e.target.value;
     this.setState(()=>({ 
         note
       }))
  }
  onAmountChange=(e)=>{
    const amount=e.target.value;
    if(amount.match(/^\d{1,}(\.\d{0,2})?$/) || !amount){
        this.setState(()=>({ 
            amount
          }))
    }
  }
  onSubmit=(e)=>{
        e.preventDefault(); 
        let error;
        if(!this.state.description.trim() || !this.state.amount){ 
            error="please enter amount and description";
            this.setState(()=>({ error }));
        }else{
          this.props.onSubmit({
              description:this.state.description,
              amount:this.state.amount,
              note:this.state.note,
              createdAt:this.state.createdAt
          })   
          this.setState(()=>({ error }));
        }
  }
  onDateChange=(createdAt)=>{
    if(createdAt){
        this.setState({createdAt})
    }
  }
  render(){
      return (
          <div> 
              {this.state.error && <p>{this.state.error}</p>}
              <form onSubmit={this.onSubmit}>
              <input 
                value={this.state.description}
                placeholder="Add a new Expense"
                type="text"
                name="add_input"
                autoFocus
                onChange={this.onDescriptionChange}
                /> 
                <input 
                value={this.state.amount}
                placeholder="amount"
                type="number"
                name="amount_input"
                onChange={this.onAmountChange}
                /> 
                <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.focused}
                onFocusChange={({focused})=> this.setState({focused}) }
                numberOfMonths={1}
                isOutsideRange={()=>false}
                displayFormat="DD/MM/YYYY"
                id="singleDatePicker_id"
                />
                <textarea
                value={this.state.note}
                placeholder="Add notes to your expenses(Optional !)"
                name="amount_input"
                onChange={this.onNoteChange}
                > 
                </textarea>
                <button> Add Expense </button>                 
              </form>
          </div>
      )
  }
}