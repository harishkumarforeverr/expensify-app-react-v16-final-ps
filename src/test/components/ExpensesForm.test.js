import React from "react";
import moment from "moment";
import {shallow} from "enzyme";
import ExpensesForm from "../../components/ExpensesForm";
import expenses from "../fixtures/expenses";

test('import ExpensesForm from "../../components/ExpensesForm"',()=>{
    const wrapper=shallow(<ExpensesForm />);
    expect(wrapper).toMatchSnapshot();
})


test('import ExpensesForm from "../../components/ExpensesForm with user passes data"',()=>{
    const wrapper=shallow(<ExpensesForm {...expenses[2]} />);
    expect(wrapper).toMatchSnapshot();
})

test('import ExpensesForm from "../../components/ExpensesForm with no data and error messgae undali"',()=>{
    const wrapper=shallow(<ExpensesForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find("form").simulate('submit',{preventDefault:()=>{}}) ;
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state("error").length).toBeGreaterThan(1);
})


test('import ExpensesForm from "../../components/ExpensesForm with no data and error messgae undali"',()=>{
    const wrapper=shallow(<ExpensesForm onSubmit={()=>{}} {...expenses[2]} />);
    wrapper.find("form").simulate('submit',{preventDefault:()=>{}}) ;
    expect(wrapper.state("error")).toBe(undefined);
})



test('import ExpensesForm from "../../components/ExpensesForm targeting the input tag description"',()=>{
    const value="harish";
    const wrapper=shallow(<ExpensesForm />);
    wrapper.find("input").at(0).simulate('change',{
        target:{
            value
        }
    }) ;
    expect(wrapper.state("description")).toBe(value);
})

test('import ExpensesForm from "../../components/ExpensesForm targeting the textarea"',()=>{
    const value="im bying the 60,000 laptop";
    const wrapper=shallow(<ExpensesForm />);
    wrapper.find("textarea").simulate('change',{
        target:{
            value
        }
    }) ;
    expect(wrapper.state("note")).toBe(value);
})



test('import ExpensesForm from "../../components/ExpensesForm targeting the input tag amount"',()=>{
    const value="123.54";
    const wrapper=shallow(<ExpensesForm />);
    wrapper.find("input").at(1).simulate('change',{
        target:{
            value
        }
    }) ;
    expect(wrapper.state("amount")).toBe(value);
})



test('import ExpensesForm from "../../components/ExpensesForm targeting the input tag amount"',()=>{
    const value="123.543";
    const wrapper=shallow(<ExpensesForm />);
    wrapper.find("input").at(1).simulate('change',{
        target:{
            value
        }
    });
    expect(wrapper.state("amount")).toBe("");//toBeFalsy();
})
 


test('import ExpensesForm from "../../components/ExpensesForm shuld call its props "',()=>{
    const onSubmitSpy=jest.fn();
    const wrapper=shallow(<ExpensesForm onSubmit={onSubmitSpy} {...expenses[2]} />);
    wrapper.find("form").simulate('submit',{preventDefault:()=>{}}) ;
    expect(wrapper.state("error")).toBe(undefined);
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description: expenses[2].description,
        amount: expenses[2].amount,
        note: expenses[2].note,
        createdAt: expenses[2].createdAt
    });

})



test('import ExpensesForm from "../../components/SingleDatePicker shuld call its props "',()=>{
    const now=moment(0);
    const wrapper=shallow(<ExpensesForm/>);
    wrapper.find("withStyles(SingleDatePicker)").prop('onDateChange')(now);
    expect(wrapper.state("createdAt")).toBe(now); 
})


test('import ExpensesForm from "../../components/SingleDatePicker onFocusChange "',()=>{
    const focused=true;
    const wrapper=shallow(<ExpensesForm/>);
    wrapper.find("withStyles(SingleDatePicker)").prop('onFocusChange')({focused});
    expect(wrapper.state("focused")).toBe(focused); 
})