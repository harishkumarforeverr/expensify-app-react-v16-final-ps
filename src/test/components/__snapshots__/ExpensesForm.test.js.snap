// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import ExpensesForm from "../../components/ExpensesForm with no data and error messgae undali" 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="add_input"
      onChange={[Function]}
      placeholder="Add a new Expense"
      type="text"
      value=""
    />
    <input
      name="amount_input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value=""
    />
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="DD/MM/YYYY"
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="singleDatePicker_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <textarea
      name="amount_input"
      onChange={[Function]}
      placeholder="Add notes to your expenses(Optional !)"
      value=""
    />
    <button>
       Add Expense 
    </button>
  </form>
</div>
`;

exports[`import ExpensesForm from "../../components/ExpensesForm with no data and error messgae undali" 2`] = `
<div>
  <p>
    please enter amount and description
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="add_input"
      onChange={[Function]}
      placeholder="Add a new Expense"
      type="text"
      value=""
    />
    <input
      name="amount_input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value=""
    />
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="DD/MM/YYYY"
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="singleDatePicker_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <textarea
      name="amount_input"
      onChange={[Function]}
      placeholder="Add notes to your expenses(Optional !)"
      value=""
    />
    <button>
       Add Expense 
    </button>
  </form>
</div>
`;

exports[`import ExpensesForm from "../../components/ExpensesForm with user passes data" 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="add_input"
      onChange={[Function]}
      placeholder="Add a new Expense"
      type="text"
      value="credit Card"
    />
    <input
      name="amount_input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value={4500}
    />
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-05T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="DD/MM/YYYY"
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="singleDatePicker_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <textarea
      name="amount_input"
      onChange={[Function]}
      placeholder="Add notes to your expenses(Optional !)"
      value=""
    />
    <button>
       Add Expense 
    </button>
  </form>
</div>
`;

exports[`import ExpensesForm from "../../components/ExpensesForm" 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="add_input"
      onChange={[Function]}
      placeholder="Add a new Expense"
      type="text"
      value=""
    />
    <input
      name="amount_input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value=""
    />
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="DD/MM/YYYY"
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={false}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="singleDatePicker_id"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
    <textarea
      name="amount_input"
      onChange={[Function]}
      placeholder="Add notes to your expenses(Optional !)"
      value=""
    />
    <button>
       Add Expense 
    </button>
  </form>
</div>
`;
